### PR TITLE
[GORA-527] Implement a data store for redis

### PR DESCRIPTION
**Pull request by cuent**
Redis is an open-source (BSD licensed), in-memory data structure store, used as a database, cache and message broker. 

Implementation of a new datastore that provides Redis compatibility for GORA as part of GSoC. 

See issue in [here](https://issues.apache.org/jira/browse/GORA-527?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel)
See reports [here](https://cwiki.apache.org/confluence/display/GORA/GORA-527%3A+Implement+a+data+store+for+REDIS+reports)

### CODE REVIEW

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisMapping.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  private int datebase;
+  private String prefix;
+  private Map<String, String> fields;
+  private Map<String, RedisType> types;
+
+  public int getDatebase() {
+    return datebase;
+  }
+
+  public void setDatebase(int datebase) {
```
Comment (madhawa-gunasekara) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisMapping.java]: Fix this typo

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+            } else {
+              map.fastRemoveAsync(mapping.getFields().get(field.name()));
+            }
+          }
+        } else {
+          for (Schema.Field field : fields) {
+            Object fieldValue = obj.get(field.pos());
+            String redisField = mapping.getFields().get(field.name());
+            RedisType redisType = mapping.getTypes().get(field.name());
+            switch (redisType) {
```
Comment (madhawa-gunasekara) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: default case is missing for unsupported data types

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+   */
+  @Override
+  public void initialize(Class<K> keyClass, Class<T> persistentClass, Properties properties) throws GoraException {
+    try {
+      super.initialize(keyClass, persistentClass, properties);
+
+      InputStream mappingStream;
+      if (properties.containsKey(XML_MAPPING_DEFINITION)) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace(XML_MAPPING_DEFINITION + " = " + properties.getProperty(XML_MAPPING_DEFINITION));
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Use parameterized logging

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+      persistent.put(field.pos(), redisVal);
+      persistent.setDirty(field.pos());
+    }
+    return countRetrieved > 0 ? persistent : null;
+  }
+
+  public T newInstanceFromHash(RMap<String, Object> map, String[] fields) throws GoraException, IOException {
+    fields = getFieldsToQuery(fields);
+    T persistent = newPersistent();
+    for (String f : fields) {
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: use more descriptive variable names

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+      RLexSortedSetAsync secundaryIndex = batchInstance.getLexSortedSet(generateIndexKey());
+      secundaryIndex.removeAsync(key.toString());
+      if (mode == StorageMode.SINGLEKEY) {
+        RMapAsync<Object, Object> map = batchInstance.getMap(generateKeyHash(key));
+        RFuture<Boolean> deleteAsync = map.deleteAsync();
+        batchInstance.execute();
+        return deleteAsync.get();
+      } else {
+        batchInstance.execute();
+        return redisInstance.getKeys().deleteByPattern(generateKeyStringBase(key) + WILDCARD) > 0;
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Why this delete operation is not included in the batch?

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+        }
+      }
+    } else {
+      for (String key : range) {
+        secundaryIndex.removeAsync(key);
+        if (mode == StorageMode.SINGLEKEY) {
+          RMapAsync<Object, Object> map = batchInstance.getMap(generateKeyHash(key));
+          map.deleteAsync();
+        } else {
+          redisInstance.getKeys().deleteByPattern(generateKeyStringBase(key) + WILDCARD);
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Why this delete operation is not included in the batch?

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    return range.size();
+  }
+
+  /**
+   * Execute the query and return the result.
+   */
+  @Override
+  public Result<K, T> execute(Query<K, T> query) throws GoraException {
+    Collection<String> range = runQuery(query);
+    RedisResult<K, T> redisResult = new RedisResult<>(this, query, range);
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Return the object directly, there is no need for this variable

File: gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+import org.apache.gora.util.AvroUtils;
+import org.apache.gora.util.IOUtils;
+import org.redisson.api.RList;
+import org.redisson.api.RMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DatumHandler<T extends PersistentBase> {
+
+  public static final Logger LOG = LoggerFactory.getLogger(DatumHandler.class);
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java]: you can use LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+  protected static final String INSTANCE_NAME_PROPERTY = "redis.instance";
+  protected static final String ZOOKEEPERS_NAME_PROPERTY = "redis.zookeepers";
+  protected static final String USERNAME_PROPERTY = "redis.user";
+  protected static final String PASSWORD_PROPERTY = "redis.password";
+  protected static final String PARSE_MAPPING_FILE_KEY = "gora.redis.mapping.file";
+  protected static final String DEFAULT_MAPPING_FILE = "gora-redis-mapping.xml";
+  protected static final String XML_MAPPING_DEFINITION = "gora.mapping";
+  private RedissonClient redisInstance;
+  private RedisMapping mapping;
+  public static final Logger LOG = LoggerFactory.getLogger(RedisStore.class);
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: You can use LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+
+  //Redis constants
+  private static final String FIELD_SEPARATOR = ".";
+  private static final String WILDCARD = "*";
+  private static final String INDEX = "index";
+  private static final String START_TAG = "{";
+  private static final String END_TAG = "}";
+  private static final String PREFIX = "redis://";
+
+  protected static final String MOCK_PROPERTY = "redis.mock";
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Are all these constants needed?

File: gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+        break;
+      default:
+        throw new AssertionError(fieldSchema.getType().name());
+    }
+    return map;
+  }
+
+  @SuppressWarnings("unchecked")
+  public List<Object> serializeFieldList(Schema fieldSchema, Object fieldValue) {
+    List<Object> sFL = new ArrayList();
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java]: use more descriptive names

File: gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+    switch (fieldSchema.getType()) {
+      case ARRAY:
+        List<?> ls = (List<?>) fieldValue;
```
Comment (carlosrmng) [inline gora-redis/src/main/java/org/apache/gora/redis/util/DatumHandler.java]: use more descriptive names

File: gora-redis/src/test/java/org/apache/gora/redis/GoraRedisTestDriver.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+ * @author Xavier Sumba
+ */
+public class GoraRedisTestDriver extends GoraTestDriver {
+
+  private static final String DOCKER_CONTAINER_NAME = "grokzen/redis-cluster:latest";
+  private FixedHostPortGenericContainer redisContainer = ((FixedHostPortGenericContainer) new FixedHostPortGenericContainer(DOCKER_CONTAINER_NAME)
+      .waitingFor(new RedisStartupLogWaitStrategy())
+      .withStartupTimeout(Duration.ofMinutes(3))
+      .withEnv("STANDALONE", "true")
+      .withEnv("SENTINEL", "true"))
```
Comment (carlosrmng) [inline gora-redis/src/test/java/org/apache/gora/redis/GoraRedisTestDriver.java]: Usually, exposing ports directly is not a good practice.

**Comment by carlosrmng** [comment]
@cuent , Please keep your PR synchronized with master. Some PR were merged and there is a conflict.

**Comment by cuent** [comment]
Thanks @madhawa-gunasekara  and @carlosrmng!  I will solve conflicts and comments by the end of this week.

**Comment by djkevincr** [comment]
@cuent Can you please take update/merge against current master ? I see so many files are added to the PR, Please add only the files where the changes are needed. Once you do that I ll have a look on your PR.

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java (hunk 5 above)
Comment (cuent) [reply gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: Por un bug estas operaciones aun no estan soportadas en batch
https://github.com/redisson/redisson/issues/2216
Vas a crear un issue para cuando se libere una nueva version de 
redisson agregar esta operaciones delete al batch.

File: gora-redis/src/test/java/org/apache/gora/redis/GoraRedisTestDriver.java (hunk 13 above)
Comment (cuent) [reply gora-redis/src/test/java/org/apache/gora/redis/GoraRedisTestDriver.java]: Unfortunately, port configuration for masters and slaves into the container is static. So, if we do a forward from a port to a random, this reference will be lost, and as a consequence, the connection won't be possible.

File: gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java (hunk 10 above)
Comment (cuent) [reply gora-redis/src/main/java/org/apache/gora/redis/store/RedisStore.java]: yes

**Comment by cuent** [comment]
> @cuent Can you please take update/merge against current master ? I see so many files are added to the PR, Please add only the files where the changes are needed. Once you do that I ll have a look on your PR.

I am not sure why those files are being added. In fact, they are showing with 0 changes, maybe it is because of a change on the hash of the branch history. I don't know how to solve that, @djkevincr  do you have any idea?

**Comment by djkevincr** [comment]
Can you try git rebase from your GORA-527 branch - git rebase master? Resolve any conflicts and update you PR?

**Comment by cuent** [comment]
files conflicts are address in a new PR. See #186

**Pull request closed without merging**